### PR TITLE
Infer sharing

### PR DESCRIPTION
**Pull request by calebmkim**
I know this code will fail the compiler tests, but this is the domination map + infer sharing stuff, in case you wanted to look at it @rachitnigam.

### CODE REVIEW

File: calyx/src/passes/infer_share.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        "Infer User Defined Components as Shareable"
+    }
+}
+
+impl ConstructVisitor for InferShare {
+    fn from(ctx: &ir::Context) -> CalyxResult<Self>
+    where
+        Self: Sized + Named,
+    {
+        let opts = Self::get_opts(&["print_dmap"], ctx);
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:61]: Use `print-dmap` instead of `print_dmap`

File: calyx/src/analysis/dominator_map.rs [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+///   @NODE_ID(6) par {
+///     @NODE_ID(7) C;
+///     @NODE_ID(8) D;
+///   }
+///   @NODE_ID(9) E;
+/// }
+///
+/// These identifiers are used by the compilation methods [calculate_states_recur]
+/// and [control_exits]
+fn compute_unique_ids(con: &mut ir::Control, mut cur_state: u64) -> u64 {
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:92]: There is some overlap between this and the stuff tdcc does. Can we open an issue about that?

File: calyx/src/analysis/dominator_map.rs [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+        ir::Control::Empty(_) => cur_state,
+    }
+}
+
+impl DominatorMap {
+    /// Construct a domination map.
+    pub fn new(control: &mut ir::Control, comp: &ir::Component) -> Self {
+        compute_unique_ids(control, 0);
+        let mut pred_map = HashMap::new();
+        Self::build_predecessor_map(control, &mut pred_map);
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:211]: This is not quite computing the predecessors so let's rename to be bit more accurate

File: calyx/src/passes/infer_share.rs [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+                    return false;
+                }
+            }
+            unreachable!("id should match either beginning or ending id")
+        }
+        _ => true,
+    }
+}
+
+pub struct InferShare {
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:37]: This needs a documentation comment (`///`) with a short explanation of what the pass is doing. It's okay to just refer to the DominatorMap logic while explaining it.

File: calyx/src/passes/infer_share.rs [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+impl ConstructVisitor for InferShare {
+    fn from(ctx: &ir::Context) -> CalyxResult<Self>
+    where
+        Self: Sized + Named,
+    {
+        let opts = Self::get_opts(&["print-dmap"], ctx);
+
+        let state_shareable = ShareSet::from_context::<true>(ctx);
+
+        let mem_ids = HashSet::from([
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:65]: Why do we explicitly need to mark memories as unshareable? What happens if other primitives are unshareable in the future? In general, it is better for the pass to be conservative and not attempt to share things that have not been explicitly marked as shareable

File: calyx/src/passes/infer_share.rs [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        {
+            if let ir::PortParent::Cell(cell_wref) = &dst_ref.parent {
+                return Some(Rc::clone(&cell_wref.upgrade()));
+            }
+        }
+        None
+    }
+
+    //returns true if port is 1'd1
+    fn src_const(port: &ir::RRC<ir::Port>) -> bool {
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:313]: This seems like an unnecessary function. Just use the `is_constant` call directly

File: calyx/src/passes/infer_share.rs [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+        port.borrow().is_constant(1, 1)
+    }
+
+    //returns true if guard is True
+    fn guard_true(guard: &ir::Guard) -> bool {
+        matches!(guard, ir::Guard::True)
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:319]: Use the `is_true` method on `ir::Guard` directly instead of defining this on this struct.

File: tests/passes/infer-share-pass/basic1.futil [hunk 8]
```diff
@@ -0,0 +1,53 @@
+//-p infer-share
```
Comment (rachitnigam) [inline tests/passes/infer-share-pass/basic1.futil:1]: Can you add a comment explaining what this test is doing and rename it to have a more informative name like "simple-dominate" or something?

File: tests/passes/infer-share-pass/basic2.futil [hunk 9]
```diff
@@ -0,0 +1,98 @@
+//-p infer-share
+
```
Comment (rachitnigam) [inline tests/passes/infer-share-pass/basic2.futil:2]: Comment explaining the test and rename to "seq-dominate" or something

File: tests/passes/infer-share-pass/invoke.futil [hunk 10]
```diff
@@ -0,0 +1,93 @@
+//-p infer-share
```
Comment (rachitnigam) [inline tests/passes/infer-share-pass/invoke.futil:1]: Comment for test and better name for file needed

File: tests/passes/infer-share-pass/invoke2.futil [hunk 11]
```diff
@@ -0,0 +1,85 @@
+//-p infer-share
```
Comment (rachitnigam) [inline tests/passes/infer-share-pass/invoke2.futil:1]: Comment for test and better name for file needed

File: tests/passes/infer-share-pass/invoke3.futil [hunk 12]
```diff
@@ -0,0 +1,73 @@
+//-p infer-share
```
Comment (rachitnigam) [inline tests/passes/infer-share-pass/invoke3.futil:1]: Comment for test and better name for file needed

File: calyx/src/passes/infer_share.rs [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+            && Self::guard_only_dones(&assignment.guard)
+    }
+
+    //returns true if port is a "done" port or is a constant
+    fn is_done_port(port: &ir::RRC<ir::Port>) -> bool {
+        port.borrow().name.id == "done" || port.borrow().is_constant(1, 1)
+    }
+
+    //returns true if guard only contains dones, or is true
+    fn guard_only_dones(guard: &ir::Guard) -> bool {
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:184]: `ir::Guard` guard has a method called `all_ports` that will make writing this function much easier. It essentially gives you all the ports used in the guard AST

File: calyx/src/passes/infer_share.rs [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+        comp.attributes.insert("state_share", 1);
+        self.state_shareable.add(comp.name.clone());
+        Ok(Action::Continue)
+    }
+}
+
+///Contains the ids of all the cells that are read from in a given "node" in
+///the domination map graph
+#[derive(Default)]
+struct ReadSet {
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:153]: If this struct is used to manage the DominatorMap then it should be not be in the `infer_share` file

File: calyx/src/passes/infer_share.rs [hunk 15]
```diff
@@ -0,0 +1,386 @@
...
+        ) {
+            if share.is_shareable_component(&cell) {
+                self.reads.insert(cell.borrow().name().clone());
+            }
+        }
+    }
+
+    //Given a control statement c, adds all of the reads of shareable cells from c.
+    //For while loops and if stmts, the control refers only to the guard, not the body.
+    fn get_reads_from_control(
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:218]: The `ReadWriteSet::control_port_read_write_set` defines a way to get all the reads and writes from a control tree

File: calyx/src/passes/infer_share.rs [hunk 16]
```diff
@@ -0,0 +1,386 @@
...
+}
+
+impl NameSearch {
+    fn new(name: ir::Id) -> Self {
+        NameSearch { name }
+    }
+
+    //returns true if the port's parent's name matches name, false otherwise
+    fn parent_matches_name(&self, port: &ir::RRC<ir::Port>) -> bool {
+        if let ir::PortParent::Cell(cell) = &port.borrow().parent {
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:288]: You can use the `PortParent::is_hole()` and `PortParent::get_parent_name` methods here instead:
```
let parent = port.borrow().parent;
return !parent.is_hole() && parent.get_parent_name() == self.name
```

File: calyx/src/passes/infer_share.rs [hunk 17]
```diff
@@ -0,0 +1,386 @@
...
+        port.borrow().is_constant(1, 1)
+    }
+
+    //returns true if guard is True
+    fn guard_true(guard: &ir::Guard) -> bool {
+        matches!(guard, ir::Guard::True)
+    }
+
+    //given a vec of assignments, return true if name is ever written to in
+    //assignments, false otherwise
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:323]: The comment doesn't seem to match what this function does–it seems to filter out all the assignments to a cell's `go` signal and checks if the `src` is the same as `self.name`. Can you try inlining the logic of `Self::guaranteed_go`, etc. into this function and renaming it to be consistent with what's going on in it?

File: calyx/src/passes/infer_share.rs [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+    //with self's name.
+    fn is_written(
+        &self,
+        dominators: &HashSet<u64>,
+        comp: &mut ir::Component,
+        shareable: &ShareSet,
+    ) -> bool {
+        for dominator in dominators {
+            if let Some(c) =
+                DominatorMap::get_control(*dominator, &comp.control.borrow())
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:343]: It seems like a common operation on DominatorMap is to get a bunch of dominators and get the associated control nodes. Instead of iterating on the control tree over and over again, can we define a `DominatorMap::get_control_nodes` method that gets all of them at once? This make it easier for the API to get a bunch of control nodes in one go.

File: calyx/src/passes/infer_share.rs [hunk 19]
```diff
@@ -0,0 +1,386 @@
...
+                        }
+                        if shareable.is_shareable_component(comp)
+                            && !inputs.is_empty()
+                        {
+                            return *comp.borrow().name() == self.name;
+                        }
+                    }
+                }
+            } else {
+                unreachable!("should always be able to get control from id")
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:381]: If this unreachable then change the `if let` into a `.get(..).unwrap_or_else(|| unreachable!(..))` call above

File: calyx/src/passes/infer_share.rs [hunk 20]
```diff
@@ -0,0 +1,386 @@
...
+    //if the assignment reads only dones, return true. This is used so that we
+    //can ignore reads of "done" cells.
+    fn reads_only_dones(assignment: &ir::Assignment) -> bool {
+        Self::is_done_port(&assignment.src)
+            && Self::guard_only_dones(&assignment.guard)
+    }
+
+    //returns true if port is a "done" port or is a constant
+    fn is_done_port(port: &ir::RRC<ir::Port>) -> bool {
+        port.borrow().name.id == "done" || port.borrow().is_constant(1, 1)
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:180]: This doesn't seem to be checking if something is a done port. Things that are not named `done` can be done ports if they have the `@done` attribute. Furthermore, the `is_constant` call not checking whether something is a done call at all

File: calyx/src/passes/infer_share.rs [hunk 21]
```diff
@@ -0,0 +1,386 @@
...
+    //given a port, insert the port's parent's id if the parent the port's parent
+    //is shareable
+    fn add_parent_if_shareable(
+        &mut self,
+        share: &ShareSet,
+        port: &ir::RRC<ir::Port>,
+    ) {
+        if let ir::PortParent::Cell(cell) = &port.borrow().parent {
+            if share.is_shareable_component(&cell.upgrade()) {
+                self.reads.insert(cell.upgrade().borrow().name().clone());
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:166]: Try importing the `CloneName` trait so that you can use the `clone_name` method instead of the `borrow().name().clone()` calls everywhere

File: calyx/src/analysis/dominator_map.rs [hunk 22]
```diff
@@ -0,0 +1,489 @@
...
+            Self::update_map(main_c, 0, &empty_set, d_map);
+        }
+    }
+
+    //Given a control, gets its associated id. For if statments, gets the
+    //beginning id if begin_id is true and end_id if begin_id is false.
+    //Should not be called on empty control
+    //statements or any other statements that don't have an id numbering.
+    fn get_id(c: &ir::Control, begin_id: bool) -> u64 {
+        if let Some(v) = match c {
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:237]: Change this into a `let v = match { ... }; v.unwrap_or_else(|| unreachable!(..))` call instead of the `if let` stuff. In general, if the `None` branch is an exception, use `unwrap_or_else`

File: calyx/src/analysis/dominator_map.rs [hunk 23]
```diff
@@ -0,0 +1,489 @@
...
+    }
+
+    //given a control stmt c and a key, returns true if c matches key, false
+    //otherwise. For if stmts return true if key matches either begin or end id.
+    fn matches_key(c: &ir::Control, key: u64) -> bool {
+        let mut ids = vec![Self::get_id(c, true)];
+        if let Some(end) = get_attr(c, END_ID) {
+            ids.push(end);
+        }
+        ids.contains(&key)
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:262]: Why construct a vector and call contains on it? It seems we'll only ever have `key == end` or `key == Self::get_id(c, true)`. Just do those comparisons directly.

File: calyx/src/analysis/dominator_map.rs [hunk 24]
```diff
@@ -0,0 +1,489 @@
...
+            og_map = d_map.map.clone();
+            Self::update_map(main_c, 0, &empty_set, d_map);
+        }
+    }
+
+    //Given a control, gets its associated id. For if statments, gets the
+    //beginning id if begin_id is true and end_id if begin_id is false.
+    //Should not be called on empty control
+    //statements or any other statements that don't have an id numbering.
+    fn get_id(c: &ir::Control, begin_id: bool) -> u64 {
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:236]: Make `begin_id` a compile time constant. All uses of `get_id` seem to pass a concrete value to the function.

File: calyx/src/analysis/dominator_map.rs [hunk 25]
```diff
@@ -0,0 +1,489 @@
...
+                }
+                None
+            }
+        }
+    }
+
+    //gets attribute s from c, panics otherwise. Should be used when you know
+    //that c has attribute s.
+    fn get_guaranteed_attribute(c: &ir::Control, s: &str) -> u64 {
+        match get_attr(c, s) {
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:311]: Use `unwrap_or_else`

File: calyx/src/analysis/dominator_map.rs [hunk 26]
```diff
@@ -0,0 +1,489 @@
...
+            exits_map,
+            component_name,
+        };
+        Self::build_map(control, &mut map);
+        map
+    }
+
+    //Builds the domination map by running update_map() until the map
+    //stops changing.
+    fn build_map(main_c: &ir::Control, d_map: &mut DominatorMap) {
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:222]: Why does this not take a `&mut self` instead of a `d_map: &mut ...`

File: calyx/src/analysis/dominator_map.rs [hunk 27]
```diff
@@ -0,0 +1,489 @@
...
+            attributes.insert(NODE_ID, cur_state);
+            cur_state += 1;
+            cur_state = compute_unique_ids(body, cur_state);
+            cur_state
+        }
+        ir::Control::Empty(_) => cur_state,
+    }
+}
+
+impl DominatorMap {
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:205]: A bunch of these functions, like `get_id` and `get_attr` don't seem to be related to a DominatorMap in particular. In that case, we should move them into the file scope and use them there. In general, functions that act upon a DominatorMap reference or construct a DominatorMap should be inside the `impl`.

File: calyx/src/analysis/dominator_map.rs [hunk 28]
```diff
@@ -0,0 +1,489 @@
...
+            }
+        }
+    }
+
+    //Gets the "final" nodes in control c. This useful for getting
+    //what will be the predecessors of the next node in the control sequence.
+    fn get_final(c: &ir::Control) -> HashSet<u64> {
+        let mut hs = HashSet::new();
+        match c {
+            ir::Control::Empty(_) => panic!("To Do: deal w/ empty controls"),
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:362]: Do we actually need to do anything for empty?

File: calyx/src/analysis/dominator_map.rs [hunk 29]
```diff
@@ -0,0 +1,489 @@
...
+            | ir::Control::Enable(_)
+            | ir::Control::While(_) => {
+                hs.insert(Self::get_guaranteed_attribute(c, NODE_ID));
+            }
+            ir::Control::If(_) => {
+                hs.insert(Self::get_guaranteed_attribute(c, END_ID));
+            }
+            ir::Control::Seq(ir::Seq { stmts, .. }) => {
+                match (&stmts[..]).last() {
+                    None => panic!("error: empty Seq block. Run ___ "),
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:373]: Run what? `crate::passes::CollapseControl::name()`?

Comment (rachitnigam) [reply calyx/src/analysis/dominator_map.rs:373]: Also, use `unwrap_or_else`

File: calyx/src/analysis/dominator_map.rs [hunk 30]
```diff
@@ -0,0 +1,489 @@
...
+                }
+            }
+        }
+        hs
+    }
+
+    //Given an id and its predecessors pred, and a domination map d_map, updates
+    //d_map accordingly (i.e. the union of all dominators of the predecessors
+    //plus itself).
+    fn update_node(pred: &HashSet<u64>, id: u64, d_map: &mut DominatorMap) {
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:390]: Why doesn't this take a `&mut self`?

File: calyx/src/analysis/dominator_map.rs [hunk 31]
```diff
@@ -0,0 +1,489 @@
...
+        union.insert(id);
+        d_map.map.insert(id, union);
+    }
+
+    //Looks through each "node" in the "graph" and updates the dominators accordingly
+    fn update_map(
+        main_c: &ir::Control,
+        cur_id: u64,
+        pred: &HashSet<u64>,
+        d_map: &mut DominatorMap,
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:406]: Same question about `&mut self`

File: calyx/src/analysis/dominator_map.rs [hunk 32]
```diff
@@ -0,0 +1,489 @@
...
+            ir::Control::Empty(_) => {
+                unreachable!(
+                    "should not pattern match agaisnt empty in update_map()"
+                )
+            }
+            ir::Control::Invoke(_) | ir::Control::Enable(_) => {
+                Self::update_node(pred, cur_id, d_map);
+            }
+            ir::Control::Seq(ir::Seq { stmts, .. }) => {
+                //Could try to think a way of doing it w/o this first stuff
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:422]: I don't think that this `first` stuff is needed. What is the problem here?

File: calyx/src/analysis/dominator_map.rs [hunk 33]
```diff
@@ -0,0 +1,489 @@
...
+                        Self::update_map(
+                            main_c,
+                            id,
+                            &d_map
+                                .exits_map
+                                .get(&prev_id)
+                                .unwrap_or_else(|| {
+                                    unreachable!(
+                                        "exit node map does not have value for {}",
+                                        prev_id
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:440]:
```suggestion
                                        "exit node map does not have value for {prev_id}",
```

File: calyx/src/analysis/dominator_map.rs [hunk 34]
```diff
@@ -0,0 +1,489 @@
...
+                    prev_id = Self::get_id(stmt, false);
+                }
+            }
+            ir::Control::Par(ir::Par { stmts, .. }) => {
+                for stmt in stmts {
+                    let id = Self::get_id(stmt, true);
+                    Self::update_map(main_c, id, pred, d_map);
+                }
+            }
+            ir::Control::While(ir::While { body, .. }) => {
```
Comment (rachitnigam) [inline calyx/src/analysis/dominator_map.rs:456]: Can you write a comment explaining why it is safe to ignore the comb group in the `while` and `if` cases?

File: calyx/src/passes/infer_share.rs [hunk 35]
```diff
@@ -0,0 +1,386 @@
...
+                let key = NameSearch::new(cell_name);
+                if !key.is_written(dominators, comp, &self.state_shareable) {
+                    self.no_share.insert(comp.name.clone());
+                    return Ok(Action::Continue);
+                }
+            }
+        }
+        comp.attributes.insert("state_share", 1);
+        self.state_shareable.add(comp.name.clone());
+        Ok(Action::Continue)
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:146]: Return `Action::Stop` if the pass does not need to traverse the control program

**Review by rachitnigam — CHANGES_REQUESTED** [review]
Left a bunch of style comments. Another high-level piece of feedback is to shrink the test suite a little bit. Ensure that each test tests a specific and small piece of functionality and that it is somewhat orthogonal to other tests.

Put the domination map tests in a folder called `tests/passes/domination-map`

File: calyx/src/analysis/dominator_map.rs (hunk 32 above)
Comment (calebmkim) [reply calyx/src/analysis/dominator_map.rs:422]: `update_map()` takes an` &HashMap` as its argument for `pred`. 
The first iteration, we want to pass `pred` as the `pred` argument when calling `update_map()` recursively. 
The subsequent iterations, we want to use `exits_map` to determine what to pass as the next `pred`. 
The only way we can use `exits_map` to determine `pred` for the first statement is if we had the id of the control statement directly preceding the `seq`.

File: calyx/src/analysis/dominator_map.rs (hunk 33 above)
Comment (calebmkim) [reply calyx/src/analysis/dominator_map.rs:440]: When I try to do this locally, it gives me the following warning: 

panic message contains an unused formatting placeholder
`#[warn(non_fmt_panics)]` on by default

Comment (calebmkim) [reply calyx/src/analysis/dominator_map.rs:440]: if prefers: `unreachable!("{}", "exit node map does not have value for {prev_id}", )`

File: calyx/src/analysis/dominator_map.rs (hunk 34 above)
Comment (calebmkim) [reply calyx/src/analysis/dominator_map.rs:456]: I'm not ignoring the comb group-- in the domination map, if I attach a NODE_ID to a while control statement, that "node" refers to only the while guard (which I include to mean its combinational group and its lt.out).

Comment (rachitnigam) [reply calyx/src/analysis/dominator_map.rs:403]: Okay, can you add a comment to that effect in the code?

File: calyx/src/analysis/dominator_map.rs (hunk 24 above)
Comment (rachitnigam) [reply calyx/src/analysis/dominator_map.rs]: This doesn't seem to be resolved

Comment (calebmkim) [reply calyx/src/analysis/dominator_map.rs]: Screen Shot 2022-07-13 at 3 51 18 PM

Is this not what you meant for me to do? Or are you not seeing the change?

Comment (rachitnigam) [reply calyx/src/analysis/dominator_map.rs]: Argh, clicking on the file link actually seems to take you to the *commit* when the comment was added and not the current version of the file. Anyways, yes, I see the change now

File: calyx/src/passes/infer_share.rs (hunk 15 above)
Comment (rachitnigam) [reply calyx/src/passes/infer_share.rs]: I've taken a shot at fixing it. See if it still works

File: calyx/src/analysis/node_analysis.rs [hunk 36]
```diff
@@ -0,0 +1,221 @@
+use crate::analysis::{DominatorMap, ReadWriteSet, ShareSet};
```
Comment (rachitnigam) [inline calyx/src/analysis/node_analysis.rs:1]: Isn't this concept of `NODES` fundamentally tied to how dominator map works? If so, we should not expose this as a separate module and just provide the struct from the dominator map stuff. If you want different files for organizing purposes, create a new folder insider `analysis/` and expose all of the stuff from `analysis/dominator_map/mod.rs` in the way `analysis/mod.rs` does.

**Comment by rachitnigam** [comment]
Awesome, this looks good! Thanks for bearing with me on the back and forth on this PR!! Hit the green button when you're ready
